pkg/server/node.go

```diff
@@ -1468,36 +1473,40 @@ func (s *lockedMuxStream) Send(e *kvpb.MuxRangeFeedEvent) error {
 	return s.wrapped.Send(e)
 }
 
-// MuxRangeFeed implements the kvpb.InternalServer interface.
+// MuxRangeFeed implements the roachpb.InternalServer interface.
 func (n *Node) MuxRangeFeed(stream kvpb.Internal_MuxRangeFeedServer) error {
-	ctx, cancelFeeds := n.stopper.WithCancelOnQuiesce(stream.Context())
-	defer cancelFeeds()
-	rfGrp := ctxgroup.WithContext(ctx)
-
 	muxStream := &lockedMuxStream{wrapped: stream}
+	var streams syncutil.IntMap
 	for {
 		req, err := stream.Recv()
 		if err != nil {
-			cancelFeeds()
-			return errors.CombineErrors(err, rfGrp.Wait())
+			return err
+		}
+
+		streamCtx := n.AnnotateCtx(stream.Context())
+		streamCtx = logtags.AddTag(streamCtx, "r", req.RangeID)
+		streamCtx = logtags.AddTag(streamCtx, "s", req.Replica.StoreID)
+
+		sink := setRangeIDEventSink{
+			ctx:      streamCtx,
+			rangeID:  req.RangeID,
+			streamID: req.StreamID,
+			wrapped:  muxStream,
 		}
 
-		rfGrp.GoCtx(func(ctx context.Context) error {
-			ctx = n.AnnotateCtx(ctx)
-			ctx = logtags.AddTag(ctx, "r", req.RangeID)
-			ctx = logtags.AddTag(ctx, "s", req.Replica.StoreID)
-			ctx, restore := pprofutil.SetProfilerLabelsFromCtxTags(ctx)
-			defer restore()
-			ctx, span := tracing.ForkSpan(ctx, "mux-rf")
-			defer span.Finish()
-
-			sink := setRangeIDEventSink{
-				ctx:      ctx,
-				rangeID:  req.RangeID,
-				streamID: req.StreamID,
-				wrapped:  muxStream,
+		f := n.stores.RangeFeedPromise(req, &sink)
+		streams.Store(req.StreamID, unsafe.Pointer(f))
+		f.WhenReady(func(err error) {
+			streams.Delete(req.StreamID)
+			if err != nil {
+				var event kvpb.RangeFeedEvent
+				event.SetValue(&kvpb.RangeFeedError{
+					Error: *kvpb.NewError(err),
+				})
+				// Sending could fail, but if it did, the stream is broken anyway, so
+				// nothing we can do with this error.
+				_ = sink.Send(&event)
```

Comment:
Should we close the sink here? That would prevent stuffing subsequent errors or events into it, and also make sure the client isn't left hanging if the send fails for a non-terminal reason.

Reply:
I don't think sink has a close method.  I could store some additional state in sink, but presumably if Send returned
an error, all subsequent sends ought to return an error as well:

```
// Send blocks until it sends m, the stream is done, or the stream breaks.
// Send must be safe to call on the same stream in different goroutines.
Send(*kvpb.RangeFeedEvent) error
```